contrib/layout.c

```diff
@@ -0,0 +1,294 @@
+	{
+		if ( i < main_amount ) /* main area. */
+		{
+			view_x      = 0;
+			view_width  = main_size;
+			view_height = height / MIN(main_amount, view_amount);
+			view_y      = i * view_height;
+		}
+		else /* Stack area. */
+		{
+			view_x      = main_size;
+			view_width  = stack_size;
+			view_height = height / ( view_amount - main_amount);
+			view_y      = (i - main_amount) * view_height;
+		}
+
+		zriver_layout_v1_set_view_dimensions(output->layout, serial, i,
+				view_x + padding, view_y + padding,
+				view_width - (2*padding), view_height - (2*padding));
+	}
+
+	/* Do you want to give users some sort of status about your layout? Then
+	 * this is the place to do it! You can provide whatever arbitrary string
+	 * your want as the layout name, which other clients, such as status
+	 * bars, may fetch and display to the user. But with this example client
+	 * we are lazy and boring, so we just set a static name.
+	 */
+	zriver_layout_v1_commit(output->layout, serial, "[]=");
+}
+
+static void layout_handle_namespace_in_use (void *data, struct zriver_layout_v1 *zriver_layout_v1)
+{
+	/* Oh no, the namespace we choose is already used by another client!
+	 * All we can do now is destroy the river_layout object. Because we are
+	 * lazy, we just abort and let our cleanup mechanism destroy it. A more
+	 * sophisticated client could instead destroy only the one single
+	 * affected river_layout object and recover from this mishap. Writing
+	 * such a client is left as an exercise for the reader.
+	 */
+	struct Context *context = (struct Context *)data;
+	fputs("Invalid namespace.\n", stderr);
+	context->run = false;
+}
+
+static void noop () { }
+
+static const struct zriver_layout_v1_listener layout_listener = {
+	.namespace_in_use = layout_handle_namespace_in_use,
+	.layout_demand    = layout_handle_layout_demand,
+	.advertise_view   = noop,
+	.advertise_done   = noop,
+};
+
+static void configure_output (struct Output *output)
+{
+	/* The namespace of the layout is how the compositor chooses what layout
+	 * to use. It can be any arbitrary string. It should describe roughly
+	 * what kind of layout your client will create, so here we use "tile".
+	 */
+	output->layout = zriver_layout_manager_v1_get_river_layout(
+			output->context->layout_manager, output->output, "tile");
+	zriver_layout_v1_add_listener(output->layout, &layout_listener, output);
+}
+
+static bool create_output (struct Context *context, struct wl_registry *registry,
+		uint32_t name, const char *interface, uint32_t version)
+{
+	struct Output *output = calloc(1, sizeof(struct Output));
+	if ( output == NULL )
+		return false;
+
+	output->output = wl_registry_bind(registry, name, &wl_output_interface, 3);
+	output->context = context;
+	output->layout = NULL;
+
+
+	/* If we already have the river_layout_manager, we can get a
+	 * river_layout for this output.
+	 */
+	if ( context->layout_manager != NULL )
+		configure_output(output);
```

Comment:
No reason to do this here, I'd just create all the objects at once after all globals have been advertised. Less code is better!

Reply:
This is needed for when outputs are added later.

Reply:
When we just create the `river_layout` objects once after the registry roundtrip, outputs added while we are in the main loop won't get a `river_layout`. We really do need to do this twice.

Reply:
Ah of course, don't mind me.